contracts/core/SequencerSoloClient-ancestorProof.sol

```
@@ -0,0 +1,153 @@
+    // Doing this in a hash map is probably nicer here.
+    //
+    error CannotUpdatePendingOptimisticConsensusState();
+    error AppHashHasNotPassedFraudProofWindow();
+    error NonMembershipProofsNotYetImplemented();
+    error NoConsensusStateAtHeight(uint256 height);
+
+    constructor(ISignatureVerifier verifier_, L1Block _l1BlockProvider) {
+        verifier = verifier_;
+        l1BlockProvider = _l1BlockProvider;
+    }
+
+    /**
+     * @inheritdoc ILightClient
+     * @param proof An array of bytes that contain the l1blockhash and the sequencer's signature. The first 32 bytes of
+     * this argument should be
+     * the l1BlockHash, and the remaining bytes should be the sequencer signature which attests to the peptide AppHash
+     * for that l1BlockHash
+     */
+    function updateClient(bytes calldata proof, uint256 peptideHeight, uint256 peptideAppHash) external override {
+
+        // Here proof will contain the signed header like in the previous one. Then we will need to pass in the  
+        // the length of block heights from 1 to others. 
+
+        console2.log("length");
+        console2.log(proof.length);
+
+        if (l1BlockProvider.hash() != bytes32(proof[:32])) {
+            revert InvalidL1Origin();
+        }
+        _updateClient(proof, peptideHeight, peptideAppHash);
+    }
+
+    /* Use if you don't need to use a checkpointed block, to save a SLOAD and 5k gas */
+    function updateClientFromCheckpoint(bytes calldata proof, uint256 peptideHeight, uint256 peptideAppHash) external {
+        if (!checkpoints[bytes32(proof[:32])]) {
+            revert InvalidL1Origin();
+        }
+        _updateClient(proof, peptideHeight, peptideAppHash);
+
+        delete checkpoints[bytes32(proof[:32])];
+    }
+
+    function _updateClient(bytes calldata proof, uint256 peptideHeight, uint256 peptideAppHash) internal {
+        if (consensusStates[peptideHeight] != 0) {
```

would this be a problem if we reorg? the contract would "know" of an apphash at height H which was later reorg'd out and a new update at the same height H but with a different apphash would be accepted with no verification

---

a tx replay after reorg would revert with InvalidOrigin. but we should revert instead of silencing errors if appHash already exists?